# 01. Importing libraries

In [36]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [37]:
path = r'C:\Users\s.wienecke\Documents\Fortbildung\Instacart Basket Analysis'

In [38]:
# df_prods stands for products.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [39]:
# df stands for order.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data Consistency Checks

In [40]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [41]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

# 04. Mixed-Type Data

In [42]:
# create a datadrame
df_test = pd.DataFrame()

In [43]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [44]:
df_test

mix
0     a
1     b
2     1
3  True

In [45]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [46]:
# change data type 
df_test['mix'] = df_test['mix'].astype('str')

# 05. Missing values

In [47]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [48]:
# subset of the dataframe containing only the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [49]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [50]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [51]:
# median function for replacing numeric values
df_prods.median()

product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [52]:
# number of columns and rows of df_ prods
df_prods.shape

(49693, 5)

In [53]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [54]:
# number of columns and rows of df_ prods after cleaning, 16 rows less that df_ prods
df_prods_clean.shape

(49677, 5)

# 06. Duplicates

In [55]:
# identifying  full duplicates—multiple rows that have the exact same values in every column
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [56]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [57]:
#check the current number of rows in your df_prods_clean
df_prods_clean.shape

(49677, 5)

In [58]:
# new dataframe that doesn’t include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [59]:
df_prods_clean_no_dups.shape

(49672, 5)

# 07. Exporting Changes

In [60]:
# export df_prods_clean_no_dups to folder Prepared Data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_clean.csv'))

# 4.5 Exercise

# 2.Task

In [61]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum of the price is with 9999$, which is much too high than the 75% percentile of 11. This value is weird and therefore it could be a mistake.

# 3. Task

In [62]:
# check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There is no column with mixed typed.

# 4. Task

no mixed-type data could be found, that's why there is nothing to do

# 5. Task

In [63]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [64]:
# showing dataframe containing only the missing values
df_ords[df_ords['days_since_prior_order'].isnull() == True]

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_time  days_since_prior_order  
0                         2           8                     NaN  
11                        2          11                     NaN  
26                        1          14                     NaN  
39                        6          11                     NaN  
45                        3          12                     NaN  
...                     ...         ...                     ...  
3420930                   4          12                     NaN  
3420934                   3          18                     NaN  
3421002                   6          19                     NaN  
3421019                   1          15                     NaN  
3421069                   3          11                     NaN  

[206209 rows x 8 columns]

the column days_since_prior_order has 206209 missing values.

# 6. Task

In [65]:
# showing number of rows with missing values
df_ords.shape

(3421083, 8)

In [66]:
# filter out missing data with a new df
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [70]:
# showing number of rows after filter out missing data
df_ords_clean.shape

(3214874, 8)

df_ords_clean has 206209 rows less than df_ords. 

# 7. Task

In [68]:
# identifying  full duplicates—multiple rows that have the exact same values in every column
df_ords_clean_no_dup = df_ords_clean[df_ords_clean.duplicated()]

In [69]:
df_ords_clean_no_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

there are no duplikates in df_ords. I decided to drop the column so that I can have a clean export.

# 8. Task

because there are no duplicates I cannot address them

# 9. Task

already export df_prods_clean_no_dups to folder Prepared Data in the tasks before.

In [71]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_orders_clean.csv'))